src/bin/Interpret.cc

```diff
@@ -668,27 +669,95 @@ void Interpret::getValue(const std::vector<ASTNode*>* terms)
     printf(")\n");
 }
 
+namespace {
+/*
+ * This is a helper class to work around the limitation of ModelValidator
+ * which does not like formal parameter with the same name as a variable already defined in the model previously.
+ */
+class NameClashResolver {
+public:
+    NameClashResolver(Logic & logic) : logic(logic) {}
+
+    void addForbiddenVar(PTRef var) {
+        assert(logic.isVar(var));
+        forbiddenVars.insert(var);
+    }
+
+    void addFunction(SymRef sym) {
+        functionsToCheck.insert(sym);
+    }
+
+    std::vector<TemplateFunction> getSafeTemplates(Model const & model) {
+        std::vector<TemplateFunction> res;
+        unsigned num = 0;
```

Comment:
It feels like by moving the counter to the top level we can avoid some traversals, since likely many functions coming from the model counter will have similarly named arguments.